### ISSUE TITLE
Does the router support multiple 'active' views that are not in a visible view port?

### ISSUE THREAD

**Issue by oneillci** [aurelia/router#42]
have a question regarding the router and/or the page lifecycle.

I have a current Silverlight application (using Prism) that has multiple top level modules (think tabs). Each module contains it's own region and consists of a large set of business screens (30+ screens per module). Each module has a landing page with a menu as an entry point to the screens for that module.

Using these module regions we can have multiple 'active' views in the app at any one time, as switching between modules doesn't actually close an open form (the shell is essentially just showing/hiding these views). Navigating back to any form presents the data in the same state it was left in. This allows me to navigate away from an open, dirty form with data in a potentially invalid state to another module without triggering a dirty validation message. Navigating back to the original module presents whichever view was most recently active in the state it was last left in.

Is multiple active pages (on different routes) a scenario can be handled with the current router or page life cycle implementation? Maybe I would need some method of persisting the current state of a viewmodel for a screen that perhaps falls outside the traditional activate and deactivate events?

Perhaps there could be additional hooks into the lifecycle that listen to navigation events and respond differently to different to different navigation types: modifying URL manually triggers dirty check, clicking in-app back button triggers dirty check, changing module does not etc.)

The attached images approximate what I'm attempting to say. http://myapp/module1/view-1 is in a dirty state, but I can still navigate to module 2 and create/destroy any of the views in module 2. Navigating back to module 1 shows view-1 in the dirty state it was left in.

![mod1-home](https://cloud.githubusercontent.com/assets/641669/6549981/9bf1c642-c671-11e4-9f2c-48866ece0fa6.png)
![mod-1-view-1](https://cloud.githubusercontent.com/assets/641669/6549980/9bf15a9a-c671-11e4-8681-34c5c4ce1a6e.png)
![mod-2-home](https://cloud.githubusercontent.com/assets/641669/6549982/9bf33eaa-c671-11e4-9c9c-95ff71411eda.png)
![mod-2-view-2](https://cloud.githubusercontent.com/assets/641669/6549983/9bf41bae-c671-11e4-99ee-b262be962e60.png)

**Comment by EisenbergEffect** [comment]
There is a way to do this today, by marking a view model as Singleton. Then the same instance will be used when you navigate back. Assuming all important state is stored in the VM, you will be good to go.

However, there is another approach that we hope to enable in the future. This involves cache settings on the `router-view` element. So, you could turn caching on and have it re-use view/view-model instances when navigating. The thought is that you might set a cache size or something like this. So, if you've got a 100 screens, you could keep the last 5 or something around. In your tab scenario, you could keep all tabs around, scoped to that router-view. When the parent view went away, they would be discarded, but while they are still on that screen, those tabs would be kept alive regardless of whether or not they are the active tab. (Note that they would be unbound when not active in order to free up resources.) That's the plan.

I'd like to hear if you think these two options aren't sufficient and how you think we might be able to make things better.

**Comment by EisenbergEffect** [comment]
FYI We had both of these options available already in Durandal from a couple of years ago. They weren't implemented as nicely as we can do in Aurelia though.

**Comment by thomas-darling** [comment]
Hi @EisenbergEffect, sorry to revive this one, but we're facing the exact same requirement, and it's not clear to me how the support for this scenario has evolved, or how we should approach it.

Basically, we have tabs at the top, and clickig them switchs between the different areas of our app - and we need the state of those areas to preserved, such that when the user switches back to a tab, it looks exactly as when they left it.

What would the recommended approach to this be? I understand that we can use singleton view models to preserve the state, but I'm curious how the caching capabilities / plans you mentioned have evolved since then?

Also, how would we go about capturing / restoring the URL of those tabs?
I'm thinking that we would ideally need the router to actually support this scenario, such that if I navigate to something like "module2/*", then the router itself would keep track of which path was last active under module2, thus actually navigating to e.g. "module2/view2", assuming that was the last active path - does that make sense?
Alternatively I guess we would need to somehow intercept all location changes, and then manually update the tab-links so they always point to the most recent URL for each area - but I'm unclear how we should go about doing that, and if that is indeed the right approach to this problem?

**Comment by EisenbergEffect** [comment]
You can build your tabs using a child router, and the url will activate the correct tab automatically. If you want to persist the view instance across navigation, then, as stated elsewhere, you could create your own router-view that does the caching. All you need to do is copy the source of our built-in router-view and then alter it to cache View instances. @cmichaelgraham has done this and cal elaborate on his implementation. It's something we hope to build in, in the future, but it's difficult to get right in a "general" way. However, it's very easy to implement in an app-specific scenario such as yours.

**Comment by cmichaelgraham** [comment]
[This guy](http://cmichaelgraham.github.io/skel-nav-esri4-vs-ts/index-release.html#/esri-globe) does it by switching out the router-view in the case when it is the globe.  try navigating away and back and notice how the globe is as it was when you left

> * [app.html](https://github.com/cmichaelgraham/skel-nav-esri4-vs-ts/blob/master/skel-nav-esri4-vs-ts/views/app.html#L6-L11)
> * [app.ts](https://github.com/cmichaelgraham/skel-nav-esri4-vs-ts/blob/master/skel-nav-esri4-vs-ts/views/app.ts#L7-L8) note: singleton
> * [esri-globe.ts](https://github.com/cmichaelgraham/skel-nav-esri4-vs-ts/blob/master/skel-nav-esri4-vs-ts/views/esri-globe.ts#L23-L28) - view model toggles app properties

i think you could generalize this to something like this sketch
![image](https://cloud.githubusercontent.com/assets/10272832/12539196/144f541c-c2aa-11e5-8536-de7211ab169d.png)

**Comment by thomas-darling** [comment]
@cmichaelgraham, thanks for the insight, I'll look into that - seems like it could work for us.
Btw, it looks like that globe example is broken in IE11 - works perfectly in Chrome though :-)

@EisenbergEffect, thanks for the update - oh, and yeah, I just realized #173 is a dublicate of this :-)

Just to finish this one though - my question about the URL stuff was not so much about having the correct tab highlighted when inside the area, but about maintaining the correct href for each tab:

When switching back to e.g. the "module2" tab, I need the location in the browser address bar to be restored to e.g. "module2/view2" (assuming that was the last view shown in that area), so that the URL always matches the content being shown - we need this so people can easily share the URL by copying it from the address bar.

The question is, how would we go about doing that? My initial thought is that we would need to manually keep the href of the tabs updated, so it always matches the current page - maybe by subscribing to the "router:navigation:success" event and updating the href's accordingly...

My concern with this approach is that we also need to update the URL in response to user selections, without triggering the router (e.g. when the user applies a filter to a search) - if we only update the tab href in response to router events, then we would miss those URL changes.

I also tried listening for the "router:navigation:processing" event, with the intention of updating the href just before leaving a view, thus always capturing the most recent URL - but turns out the URL in the address bar is already changed by the time this event fires.

Am I missing some obvious or clever way to do this? :-)
